src/USER-PHONON/dynamical_matrix.cpp

```diff
@@ -0,0 +1,512 @@
+                                //if (screen) fprintf(screen,"%f with %f\n",dyn_element[j*3+beta],imass);
+                                //dynmat entries are set based on global atom index
+                                dynmat[3 * id + beta - 3][3 * aid[j] + alpha - 3] =
+                                        conversion * dyn_element[j*3+beta];
+                            }
+                    }
+                    //move atom back
+                    if (me == proc) x[i][alpha] += del;
+                }
+            }
+        }
+    }
+
+    memory->create(final_dynmat,int(dynlen),int(dynlen),"dynamic_matrix_buffer:buf");
+    for (int i = 0; i < dynlen; i++)
+        MPI_Reduce(dynmat[i], final_dynmat[i], int(dynlen), MPI_DOUBLE, MPI_SUM, 0, world);
+
+    delete [] dyn_element;
+
+    if (screen && me ==0 ) fprintf(screen,"Finished Calculating Dynamical Matrix\n");
+}
+
+/* ----------------------------------------------------------------------
+   write dynamical matrix
+------------------------------------------------------------------------- */
+
+void DynamicalMatrix::writeMatrix()
+{
+    // print file comment lines
+    if (!binaryflag && fp) {
+        clearerr(fp);
+        for (int i = 0; i < dynlen; i++) {
+            for (int j = 0; j < dynlen; j++) {
+                if ((j+1)%3==0) fprintf(fp, "%4.8f\n", final_dynmat[j][i]);
+                else fprintf(fp, "%4.8f ",final_dynmat[j][i]);
+            }
+        }
+    }
+    if (ferror(fp))
+        error->one(FLERR,"Error writing to file");
+
+    if (binaryflag && fp) {
+        clearerr(fp);
+        fwrite(&final_dynmat[0], sizeof(double), dynlen * dynlen, fp);
+        if (ferror(fp))
+            error->one(FLERR, "Error writing to binary file");
+    }
+}
+/* ----------------------------------------------------------------------
+   evaluate potential energy and forces
+   may migrate atoms due to reneighboring
+   return new energy, which should include nextra_global dof
+   return negative gradient stored in atom->f
+   return negative gradient for nextra_global dof in fextra
+------------------------------------------------------------------------- */
+
+void DynamicalMatrix::energy_force(int resetflag)
+{
+    // check for reneighboring
+    // always communicate since atoms move
+    int nflag = neighbor->decide();
+
+    if (nflag == 0) {
+        timer->stamp();
+        comm->forward_comm();
+        timer->stamp(Timer::COMM);
+    } else {
+        if (triclinic) domain->x2lamda(atom->nlocal);
+        domain->pbc();
+        if (domain->box_change) {
+            domain->reset_box();
+            comm->setup();
+            if (neighbor->style) neighbor->setup_bins();
+        }
+        timer->stamp();
+        comm->borders();
+        if (triclinic) domain->lamda2x(atom->nlocal+atom->nghost);
+        timer->stamp(Timer::COMM);
+        neighbor->build(1);
+        timer->stamp(Timer::NEIGH);
+    }
+    force_clear();
```

Comment:
FYI, anything above `force_clear()` should be removed from this function and only called once **before** starting the computation of the dynamical matrix with my suggested approach. the forward communication would not be needed due to using the direct displacement function and you must not (re-)build neighborlists anyway, since those can move (displaced) atoms between subdomains and change the order of atoms in the local arrays. i would also skip the `nflag == 0` branch (this is an optimization for MD runs) and simply do one neighbor list build. for as long as the displacements are smaller than half the neighbor list skin distance, there is no problem and no need to rebuild neighbor lists.